### PR TITLE
switch to utf8proc's upper/lowercase functions

### LINKED ISSUE #11471: uppercase/lowercase functions are not portable?
Since @ScottPJones was mentioning upper/lowercase functions recently, I took a quick look at them and I noticed that we are calling `towupper` and `towlower`, which are C99 functions that accept `wchar_t` arguments.

Unfortunately, this means that they are broken on Windows (where `wchar_t` is 16 bits) for any character outside the BMP.   Even on other platforms with a 32-bit `wchar_t`, they are going to return different results on different systems, and many systems will have out-of-date Unicode tables.  They are also locale-dependent; I'm not sure if this is desirable for us.

utf8proc has up-to-date upper/lower/titlecase mapping data already in its "database" (generated from http://www.unicode.org/Public/UNIDATA/UnicodeData.txt), so maybe we should just add a `utf8proc_toupper` function (etc.) to utf8proc to make this accessible.  Then we could call that (probably plus a check for the common case of ASCII codepoints).

### PR DESCRIPTION
**Pull request by stevengj**
This PR uses utf8proc rather than the system's `towlower` and `towupper`.  (Fixes #11471)  Rationale:
- Correct on Windows (outside of the BMP, the Windows functions are inevitably wrong since they use a 16-bit `wchar_t`).
- Portable — OS-independent results.
- Up-to-date with Unicode 7.  (On the latest MacOS X, the system functions are wrong for about 2000 codepoints.)

The results are also locale-independent, so they don't address the request in #7848 for more sophisticated, context- and locale-dependent results.  (I tend to agree with Jeff: that belongs in an external package like ICU.)

(@ScottPJones wants to rewrite the relevant portions of utf8proc in pure Julia.  That may take a while, though, and I don't think it makes sense to wait on that to fix this when the data is already trivially available in utf8proc.)

(While I agree with @JeffBezanson that many non-ASCII uses of lower/uppercase are probably mistakes from people who should really be using normalization+casefolding, I agree with @StefanKarpinski that `uppercase` and `lowercase` are simply functions that people expect to have in any string library, and we might as well implement them correctly.)

### CODE REVIEW

**Comment by stevengj** [comment]
As a side benefit, the new functions seem to be considerably faster on my machine (MacOS).  For
```jl
function foo(c, n)
    for i = 1:n
        c = uppercase(c)
    end
end
```
I get
```
julia> @time foo('a', 10^7)
  42.339 milliseconds (5 allocations: 160 bytes)

julia> @time foo('α', 10^7)
  82.697 milliseconds (5 allocations: 160 bytes)
```
with the new version and
```
julia> @time foo('a', 10^7)
  48.216 milliseconds (5 allocations: 160 bytes)

julia> @time foo('α', 10^7)
 226.551 milliseconds (5 allocations: 160 bytes)
```
with the old (libc) version.

**Comment by ScottPJones** [comment]
:+1:!

**Comment by ScottPJones** [comment]
I didn't mean to imply when I said that about rewriting to not use `utf8proc` that fixing this issue using `utf8proc` was not a very good thing... just didn't want you to get bothered if later on something I do supersedes this... this also has the benefit of being easily back portable to v0.3.x, n'est pas?

File: base/utf8proc.jl [hunk 1]
```diff
@@ -121,6 +121,12 @@ end
 
 charwidth(c::Char) = Int(ccall(:utf8proc_charwidth, Cint, (UInt32,), c))
 
+# faster x+y that does no overflow checking
+p(x::Char, y::UInt32) = reinterpret(Char, reinterpret(UInt32, x) + y)
```
Comment (tkelman) [inline base/utf8proc.jl]: Can this please not be a 1-letter function name? I know it's not exported, but still...

Comment (ScottPJones) [inline base/utf8proc.jl]: yep... I agree with @tkelman, no 1-letter (or 1 letterlike number such as 𝟙 :grinning:!) function names...
Also... this function wouldn't have been necessary if you all had agreed to my PR #11103! :grinning:  Maybe you could reopen #11103, and make this a lot cleaner?

**Comment by tkelman** [comment]
> this also has the benefit of being easily back portable to v0.3.x, n'est pas?

Sort of but it's not quite a direct cherry-pick, since release-0.3 is still using the last PSG release, 1.1.6, of utf8proc. Not sure whether upgrading the dependency library would change any user-visible code results, aside from fixing this particular bug.

**Comment by ScottPJones** [comment]
Ah... didn't know that... (I never use v0.3.x [except occasionally to see if something is a regression])
I always have lived on the bleeding edge!  (like using a new language, with major changes occurring, to base my future livelihood on! :grinning:)

**Comment by ScottPJones** [comment]
Could 0.3.9 be made to use a patch (?) (maintenance release?) of utf8proc?, i.e. just 1.1.6 plus @stevengj's changes?

**Comment by tkelman** [comment]
No, because I'll be tagging 0.3.9 in an hour or two. Upgrading for 0.3.10 might be a possibility but would want to test thoroughly. Completely updating the unicode tables might be too drastic for a patch release.

**Comment by ScottPJones** [comment]
Ah... OK, didn't realize that 0.3.9 was already ready, should have said 0.3.next ... and didn't realize that it would involve new Unicode tables... never mind then (I'm not going to use 0.3.x anyway :grinning:)

**Comment by stevengj** [comment]
@ScottPJones, I suspect that the right thing to do is to just change `Char - Char` and `Char ± Int32` so that they do not check for overflow by default, consistent with Julia's treatment of fixed-width integer arithmetic and its typical assumption that character data is valid except when conversions are performed.  However, that should be a separate PR.

**Comment by ScottPJones** [comment]
@stevengj I think `Char ± Int32` is incorrect now for another reason... it should return an Integer, not a `Char`, so that when/if we have `Char` always be validated, it cannot be used to bypass the abstraction...

**Comment by tkelman** [comment]
This needs to update the checksums via `NO_GIT`. It turns out the branch-tracking behavior of @vtjnash's git-externals is a bad idea, ref https://github.com/JuliaLang/julia/pull/10743#issuecomment-107050253 - we should point to tags, or strictly sha's. Note the incredibly easy to miss "SHA mismatch" warning? The broken build change on master of utf8proc breaks the Windows build here. I'm a little surprised that it doesn't break Travis too.

**Comment by stevengj** [comment]
@tkelman, could you elaborate?  I thought the `utf8proc.version` file was just a branch name and a commit hash.  What else do I need to do?

**Comment by vtjnash** [comment]
It was not intended for the tracked brach to be named "master" or "develop". It should either be a stable release brach or (typically) a tag.

**Comment by stevengj** [comment]
I see, so the SHA is not actually a commit in that branch to use, it is just a redundant checksum; you always check out the tip of that branch?  Hopefully it is fixed now.

**Comment by stevengj** [comment]
Yay, tests passing.

**Comment by ScottPJones** [comment]
Will this also make it pick up my bug fixes in `utf8proc`?
Good change, even if not

**Comment by stevengj** [comment]
@ScottPJones, yes, it's updating to the latest utf8proc version, which includes your fix for noncharacters.

**Comment by ScottPJones** [comment]
Thanks!

**Comment by tkelman** [comment]
You need to do `make -C deps install-utf8proc NO_GIT=1` to generate new checksums. Should also delete the old checksums for 1.2.

**Comment by stevengj** [comment]
Grr, why is the process of updating the dependency so annoying?

**Comment by ScottPJones** [comment]
Hehe... I'm glad *you* are working out the kinks of getting utf8proc updated, instead of me!

**Comment by tkelman** [comment]
Probably because we're stretching gmake a bit further than it should really be taken and rolling our own thing here. Which we also need to document better, and fix up so the sha is the driver rather than the branch.

**Comment by stevengj** [comment]
Okay, updated the checksums; this is the first time I've noticed the `deps/checksums` directory.

**Comment by tkelman** [comment]
Guess we didn't advertise it much, but it's been used for quite a few months now to verify integrity of downloaded tarballs.

Looks good to merge, assuming green CI.

**Comment by ScottPJones** [comment]
Yeah!

**Pull request merged**